File: OximyMac/Models/APIModels.swift
import Foundation

// MARK: - Device Registration

struct DeviceRegistrationRequest: Encodable {
    let hostname: String
    let displayName: String?
    let os: String = "mac"
    let osVersion: String
    let sensorVersion: String
    let hardwareId: String
    let ownerEmail: String?
    let permissions: Permissions

    struct Permissions: Encodable {
        let networkCapture: Bool
        let systemExtension: Bool
        let fullDiskAccess: Bool
    }
}

struct DeviceRegistrationResponse: Decodable {
    let success: Bool
    let data: DeviceData?
    let error: APIErrorData?
    let meta: ResponseMeta?

    struct DeviceData: Decodable {
        let deviceId: String
        let deviceName: String
        let deviceToken: String
        let workspaceId: String
        let config: DeviceConfig
    }
}

// MARK: - Device Config

struct DeviceConfig: Codable {
    let heartbeatIntervalSeconds: Int
    let eventBatchSize: Int
    let eventFlushIntervalSeconds: Int
    let apiEndpoint: String

    static var `default`: DeviceConfig {
        DeviceConfig(
            heartbeatIntervalSeconds: 60,
            eventBatchSize: 100,
            eventFlushIntervalSeconds: 5,
            apiEndpoint: Constants.defaultAPIEndpoint
        )
    }

    func save() {
        let defaults = UserDefaults.standard
        defaults.set(heartbeatIntervalSeconds, forKey: Constants.Defaults.heartbeatInterval)
        defaults.set(eventBatchSize, forKey: Constants.Defaults.eventBatchSize)
        defaults.set(eventFlushIntervalSeconds, forKey: Constants.Defaults.eventFlushInterval)
        defaults.set(apiEndpoint, forKey: Constants.Defaults.apiEndpoint)
    }

    static func load() -> DeviceConfig {
        let defaults = UserDefaults.standard
        return DeviceConfig(
            heartbeatIntervalSeconds: defaults.integer(forKey: Constants.Defaults.heartbeatInterval).nonZero ?? 60,
            eventBatchSize: defaults.integer(forKey: Constants.Defaults.eventBatchSize).nonZero ?? 100,
            eventFlushIntervalSeconds: defaults.integer(forKey: Constants.Defaults.eventFlushInterval).nonZero ?? 5,
            apiEndpoint: defaults.string(forKey: Constants.Defaults.apiEndpoint) ?? Constants.defaultAPIEndpoint
        )
    }
}

// MARK: - Heartbeat

struct HeartbeatRequest: Encodable {
    let sensorVersion: String
    let uptimeSeconds: Int
    let permissions: DeviceRegistrationRequest.Permissions
    let metrics: Metrics?

    struct Metrics: Encodable {
        let cpuPercent: Double?
        let memoryMb: Int?
        let eventsQueued: Int
    }
}

struct HeartbeatResponse: Decodable {
    let success: Bool
    let data: HeartbeatData?
    let error: APIErrorData?
    let meta: ResponseMeta?

    struct HeartbeatData: Decodable {
        let status: String
        let configUpdate: DeviceConfig?
        let commands: [String]?
    }
}

// MARK: - Events

struct EventBatchRequest: Encodable {
    let events: [JSONValue]
}

struct EventBatchResponse: Decodable {
    let success: Bool
    let data: EventData?
    let error: APIErrorData?
    let meta: ResponseMeta?

    struct EventData: Decodable {
        let accepted: Int
        let rejected: Int
        let deviceId: String
    }
}

// MARK: - Common

struct APIErrorData: Decodable {
    let code: String
    let message: String
}

struct ResponseMeta: Decodable {
    let requestId: String
    let timestamp: String
}

// MARK: - API Error

enum APIError: LocalizedError {
    case networkUnavailable
    case invalidEnrollmentCode
    case enrollmentExpired
    case unauthorized
    case deviceNotFound
    case conflict(String)
    case serverError(Int, String)
    case encodingError
    case decodingError(String)
    case rateLimited(retryAfter: Int?)

    var errorDescription: String? {
        switch self {
        case .networkUnavailable:
            return "No internet connection"
        case .invalidEnrollmentCode:
            return "Invalid or expired enrollment code"
        case .enrollmentExpired:
            return "Enrollment code has expired"
        case .unauthorized:
            return "Session expired. Please reconnect."
        case .deviceNotFound:
            return "Device not found. It may have been removed."
        case .conflict(let msg):
            return "Device already registered: \(msg)"
        case .serverError(let code, let msg):
            return "Server error (\(code)): \(msg)"
        case .encodingError:
            return "Failed to encode request"
        case .decodingError(let detail):
            return "Failed to decode response: \(detail)"
        case .rateLimited(let retry):
            if let retry = retry {
                return "Rate limited. Retry in \(retry) seconds."
            }
            return "Rate limited. Please try again later."
        }
    }

    static func from(_ errorData: APIErrorData?) -> APIError {
        guard let error = errorData else {
            return .serverError(0, "Unknown error")
        }
        switch error.code {
        case "bad_request":
            if error.message.lowercased().contains("expired") {
                return .enrollmentExpired
            }
            return .invalidEnrollmentCode
        case "unauthorized":
            return .unauthorized
        case "not_found":
            return .deviceNotFound
        case "conflict":
            return .conflict(error.message)
        default:
            return .serverError(0, error.message)
        }
    }
}

// MARK: - Notification Names

extension Notification.Name {
    static let authenticationFailed = Notification.Name("authenticationFailed")
}

// MARK: - JSONValue (for passing raw JSONL events)

enum JSONValue: Codable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)
    case object([String: JSONValue])
    case array([JSONValue])
    case null

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let bool = try? container.decode(Bool.self) {
            self = .bool(bool)
        } else if let int = try? container.decode(Int.self) {
            self = .int(int)
        } else if let double = try? container.decode(Double.self) {
            self = .double(double)
        } else if let string = try? container.decode(String.self) {
            self = .string(string)
        } else if let array = try? container.decode([JSONValue].self) {
            self = .array(array)
        } else if let object = try? container.decode([String: JSONValue].self) {
            self = .object(object)
        } else {
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "Invalid JSON value")
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let value):
            try container.encode(value)
        case .int(let value):
            try container.encode(value)
        case .double(let value):
            try container.encode(value)
        case .bool(let value):
            try container.encode(value)
        case .object(let value):
            try container.encode(value)
        case .array(let value):
            try container.encode(value)
        case .null:
            try container.encodeNil()
        }
    }

    init(from any: Any) {
        switch any {
        case let string as String:
            self = .string(string)
        case let int as Int:
            self = .int(int)
        case let double as Double:
            self = .double(double)
        case let bool as Bool:
            self = .bool(bool)
        case let array as [Any]:
            self = .array(array.map { JSONValue(from: $0) })
        case let dict as [String: Any]:
            self = .object(dict.mapValues { JSONValue(from: $0) })
        default:
            self = .null
        }
    }
}

// MARK: - Int Extension

extension Int {
    var nonZero: Int? {
        self == 0 ? nil : self
    }
}

// MARK: - Date Extension

extension Date {
    var relativeFormatted: String {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .abbreviated
        return formatter.localizedString(for: self, relativeTo: Date())
    }
}

File: OximyWindows/src/OximyWindows/Models/APIModels.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace OximyWindows.Models;

#region Device Registration

/// <summary>
/// Request payload for device registration.
/// </summary>
public class DeviceRegistrationRequest
{
    [JsonPropertyName("hostname")]
    public required string Hostname { get; set; }

    [JsonPropertyName("displayName")]
    public string? DisplayName { get; set; }

    [JsonPropertyName("os")]
    public string Os { get; set; } = "windows";

    [JsonPropertyName("osVersion")]
    public required string OsVersion { get; set; }

    [JsonPropertyName("sensorVersion")]
    public required string SensorVersion { get; set; }

    [JsonPropertyName("hardwareId")]
    public required string HardwareId { get; set; }

    [JsonPropertyName("permissions")]
    public DevicePermissions Permissions { get; set; } = new();
}

/// <summary>
/// Device permissions status.
/// </summary>
public class DevicePermissions
{
    [JsonPropertyName("networkCapture")]
    public bool NetworkCapture { get; set; } = true;

    [JsonPropertyName("systemExtension")]
    public bool SystemExtension { get; set; } = false;

    [JsonPropertyName("fullDiskAccess")]
    public bool FullDiskAccess { get; set; } = false;
}

/// <summary>
/// Response wrapper from device registration API.
/// </summary>
public class DeviceRegistrationResponse
{
    [JsonPropertyName("success")]
    public bool Success { get; set; }

    [JsonPropertyName("data")]
    public DeviceRegistrationData? Data { get; set; }

    [JsonPropertyName("error")]
    public ApiError? Error { get; set; }

    // Convenience accessors
    public string DeviceId => Data?.DeviceId ?? "";
    public string DeviceToken => Data?.DeviceToken ?? "";
    public string WorkspaceName => Data?.WorkspaceName ?? "";
    public string WorkspaceId => Data?.WorkspaceId ?? "";
    public DeviceConfig? Config => Data?.Config;
}

/// <summary>
/// Device registration data inside the response.
/// </summary>
public class DeviceRegistrationData
{
    [JsonPropertyName("deviceId")]
    public required string DeviceId { get; set; }

    [JsonPropertyName("deviceName")]
    public string? DeviceName { get; set; }

    [JsonPropertyName("deviceToken")]
    public required string DeviceToken { get; set; }

    [JsonPropertyName("workspaceName")]
    public string? WorkspaceName { get; set; }

    [JsonPropertyName("workspaceId")]
    public required string WorkspaceId { get; set; }

    [JsonPropertyName("config")]
    public DeviceConfig? Config { get; set; }
}

/// <summary>
/// Device configuration from server.
/// </summary>
public class DeviceConfig
{
    [JsonPropertyName("heartbeatIntervalSeconds")]
    public int HeartbeatIntervalSeconds { get; set; } = 60;

    [JsonPropertyName("eventBatchSize")]
    public int EventBatchSize { get; set; } = 100;

    [JsonPropertyName("eventFlushIntervalSeconds")]
    public int EventFlushIntervalSeconds { get; set; } = 5;

    [JsonPropertyName("apiEndpoint")]
    public string? ApiEndpoint { get; set; }
}

#endregion

#region Heartbeat

/// <summary>
/// Request payload for device heartbeat.
/// </summary>
public class HeartbeatRequest
{
    [JsonPropertyName("sensorVersion")]
    public required string SensorVersion { get; set; }

    [JsonPropertyName("uptimeSeconds")]
    public int UptimeSeconds { get; set; }

    [JsonPropertyName("permissions")]
    public DevicePermissions Permissions { get; set; } = new();

    [JsonPropertyName("metrics")]
    public DeviceMetrics Metrics { get; set; } = new();

    [JsonPropertyName("commandResults")]
    public Dictionary<string, CommandResult>? CommandResults { get; set; }
}

/// <summary>
/// Result of a command executed by the Python addon.
/// Read from ~/.oximy/command-results.json and included in next heartbeat.
/// </summary>
public class CommandResult
{
    [JsonPropertyName("success")]
    public bool Success { get; set; }

    [JsonPropertyName("executedAt")]
    public string ExecutedAt { get; set; } = "";

    [JsonPropertyName("error")]
    public string? Error { get; set; }
}

/// <summary>
/// Device metrics for heartbeat.
/// </summary>
public class DeviceMetrics
{
    [JsonPropertyName("cpuPercent")]
    public double? CpuPercent { get; set; }

    [JsonPropertyName("memoryMb")]
    public int? MemoryMb { get; set; }

    [JsonPropertyName("eventsQueued")]
    public int EventsQueued { get; set; }
}

/// <summary>
/// Response from heartbeat.
/// </summary>
public class HeartbeatResponse
{
    [JsonPropertyName("status")]
    public string Status { get; set; } = "ok";

    [JsonPropertyName("workspaceName")]
    public string? WorkspaceName { get; set; }

    [JsonPropertyName("config")]
    public DeviceConfig? Config { get; set; }

    [JsonPropertyName("commands")]
    public List<ServerCommand>? Commands { get; set; }
}

/// <summary>
/// Server command to execute.
/// </summary>
public class ServerCommand
{
    [JsonPropertyName("type")]
    public required string Type { get; set; }

    [JsonPropertyName("payload")]
    public JsonElement? Payload { get; set; }
}

#endregion

#region Events

/// <summary>
/// Request payload for event batch submission.
/// </summary>
public class EventBatchRequest
{
    [JsonPropertyName("events")]
    public required List<JsonElement> Events { get; set; }
}

/// <summary>
/// Response from event batch submission.
/// </summary>
public class EventBatchResponse
{
    [JsonPropertyName("received")]
    public int Received { get; set; }

    [JsonPropertyName("status")]
    public string Status { get; set; } = "ok";
}

#endregion

#region Sync State

/// <summary>
/// Tracks sync progress across trace files.
/// </summary>
public class SyncState
{
    [JsonPropertyName("files")]
    public Dictionary<string, FileSyncState> Files { get; set; } = new();
}

/// <summary>
/// Sync state for a single trace file.
/// </summary>
public class FileSyncState
{
    [JsonPropertyName("lastSyncedLine")]
    public int LastSyncedLine { get; set; }

    [JsonPropertyName("lastSyncedEventId")]
    public string? LastSyncedEventId { get; set; }

    [JsonPropertyName("lastSyncTime")]
    public DateTime LastSyncTime { get; set; }
}

#endregion

#region Error Response

/// <summary>
/// API error response.
/// </summary>
public class ApiErrorResponse
{
    [JsonPropertyName("error")]
    public ApiError? Error { get; set; }

    // Some APIs return message at root level
    [JsonPropertyName("message")]
    public string? Message { get; set; }

    [JsonPropertyName("statusCode")]
    public int? StatusCode { get; set; }
}

/// <summary>
/// API error details.
/// </summary>
public class ApiError
{
    [JsonPropertyName("code")]
    public string? Code { get; set; }

    [JsonPropertyName("message")]
    public string? Message { get; set; }
}

#endregion
